Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -194,24 +249,38 @@ void FixedUpdate()
             );
         }
 
-        // c. Encourage head height. //Should normalize to ~1
-        var headHeightOverFeetReward =
-            ((head.position.y - footL.position.y) + (head.position.y - footR.position.y) / 10);
-        if (float.IsNaN(headHeightOverFeetReward))
+        AddReward(matchSpeedReward * lookAtTargetReward);
+    }
+
+    //Returns the average velocity of all of the body parts
+    //Using the velocity of the hips only has shown to result in more erratic movement from the limbs, so...
+    //...using the average helps prevent this erratic movment
+    Vector3 GetAvgVelocity()
+    {
+        Vector3 velSum = Vector3.zero;
+        Vector3 avgVel = Vector3.zero;
+
+        //ALL RBS
+        int numOfRB = 0;
+        foreach (var item in m_JdController.bodyPartsList)
         {
-            throw new ArgumentException(
-                "NaN in headHeightOverFeetReward.\n" +
-                $" head.position: {head.position}\n"+
-                $" footL.position: {footL.position}\n"+
-                $" footR.position: {footR.position}"
-            );
+            numOfRB++;
+            velSum += item.rb.velocity;
         }
 
-        AddReward(
-            + 0.02f * moveTowardsTargetReward
-            + 0.02f * lookAtTargetReward
-            + 0.005f * headHeightOverFeetReward
-        );
+        avgVel = velSum / numOfRB;
+        return avgVel;
+    }
+
+    //normalized value of the difference in avg speed vs goal walking speed.
+    public float GetMatchingVelocityInverseLerp(Vector3 velocityGoal, Vector3 actualVelocity)
```

Comment:
The name of this method does not describe very well what it does.
If this is an observation, would it make more sense to return Vector3.Distance(actualVelocity, velocityGoal) directly rather than transform it (and lose information)

Reply:
yeah not sure what to do here. the returned value is both observed and used for rewards. I'm fine with just observing the Vector3.Distance but i would want to normalize that to be between (0,1) anyways. I do agree with what you're saying about losing that information which seems like a problem. I'll do a round of training using your suggestion and then post a follow up. i think you're right.